File: scripts/license-check.swift
import Foundation

/// Runs minted licence keys through the real verifier.
///
/// `scripts/lib/license.test.mjs` is JavaScript checking JavaScript: it proves
/// the issuer is self-consistent and proves nothing about the program that has
/// to accept its output. The disagreement that would actually cost money lives
/// between the two languages — base64url padding, or a signature taken over the
/// parsed claims on one side and the encoded payload on the other — and it is
/// invisible until a paying customer pastes a key that Node called valid.
///
/// So this compiles `License.swift` itself, unmodified, and feeds it cases from
/// `scripts/license-check.mjs` on stdin. A standalone `swiftc` binary rather
/// than an XCTest bundle for the same reason as `wiring-check.swift`: the Xcode
/// project has no test target, and adding one means hand-editing project.pbxproj.
///
/// Run with `make license-check`.
/// Stands in for AppInfo.swift, which `License.swift` reaches only through the
/// `major:` default — and which would drag `AppSupport` and the real bundle in
/// behind it. Every case below passes its major explicitly, so this is never the
/// value under test. `Revocations.swift` is compiled for real instead of stubbed:
/// it is four lines and no dependencies, so there is nothing to gain by faking it.
enum AppInfo {
  static let major = 1
}

@main
struct LicenseRoundTrip {
  struct Case: Decodable {
    let label: String
    let key: String
    let major: Int
    let expect: Bool
    /// Substring the refusal has to contain, when one is expected. Checked
    /// because "refused" is not the assertion that matters — a key refused for
    /// the wrong reason is a support reply that sends someone the wrong way.
    let reason: String?
  }

  static func main() {
    let input = FileHandle.standardInput.readDataToEndOfFile()
    guard let cases = try? JSONDecoder().decode([Case].self, from: input), !cases.isEmpty else {
      print("no cases on stdin — run this through `make license-check`")
      exit(2)
    }

    var failures = 0
    func pass(_ label: String) { print("  \u{1B}[32mok\u{1B}[0m   \(label)") }
    func fail(_ label: String, _ detail: String) {
      print("  \u{1B}[31mFAIL\u{1B}[0m \(label)")
      print("         \(detail)")
      failures += 1
    }

    for item in cases {
      switch LicenseKey.check(item.key, major: item.major, revoked: []) {
      case .valid(let license):
        if item.expect {
          pass("\(item.label) — \(license.email), \(license.major).x")
        } else {
          fail(item.label, "accepted a key that should have been refused")
        }
      case .refused(let reason):
        if item.expect {
          fail(item.label, "refused a genuine key: \(reason)")
        } else if let expected = item.reason, !reason.contains(expected) {
          fail(item.label, "refused for '\(reason)', expected '\(expected)'")
        } else {
          pass("\(item.label) — \(reason)")
        }
      }
    }

    print("")
    print(failures == 0 ? "  \(cases.count) checks, all passing" : "  \(failures) FAILED")
    exit(failures == 0 ? 0 : 1)
  }
}

File: scripts/license-check.mjs
#!/usr/bin/env node
// The cases `scripts/license-check.swift` runs through the real verifier.
//
// Emits JSON on stdout and nothing else, so `make license-check` can pipe it
// straight in. Every key here is minted for the check and thrown away; none is
// ever issued to anyone.
//
// The signing key is the real one, from `.env`, because that is the point:
// verifying a key signed by a throwaway keypair would prove the two languages
// agree about an algorithm while saying nothing about whether the public half
// compiled into the app matches the private half sitting on this machine. That
// mismatch is the one that refuses every key ever sold.
//
//   node --env-file-if-exists=.env scripts/license-check.mjs

import { KEY_PREFIX, mint, parse, PUBLIC_KEY, publicKeyOf } from "./lib/license.mjs";

const privateKey = process.env.LICENSE_SIGNING_KEY;
if (!privateKey) {
  console.error("FATAL: set LICENSE_SIGNING_KEY in .env (see scripts/mint-license.mjs --keygen)");
  process.exit(2);
}
if (publicKeyOf(privateKey) !== PUBLIC_KEY) {
  console.error("FATAL: LICENSE_SIGNING_KEY does not match PUBLIC_KEY in scripts/lib/license.mjs.");
  console.error("Every key signed with it would be refused by the app. Fix before issuing any.");
  process.exit(2);
}

const genuine = mint({ email: "buyer@example.com", major: 1, privateKey });

/** Re-encode claims without re-signing — what forging a key actually looks like. */
const tamper = (key, changes) => {
  const { claims, signature } = parse(key);
  const payload = Buffer.from(JSON.stringify({ ...claims, ...changes })).toString("base64url");
  return `${KEY_PREFIX}.${payload}.${signature}`;
};

console.log(
  JSON.stringify([
    { label: "a genuine key", key: genuine, major: 1, expect: true },
    {
      label: "the same key with whitespace around it",
      key: `\n  ${genuine}  \n`,
      major: 1,
      expect: true,
    },
    {
      label: "a forged email",
      key: tamper(genuine, { email: "thief@example.com" }),
      major: 1,
      expect: false,
      reason: "signature does not match",
    },
    {
      label: "a forged major",
      key: tamper(genuine, { major: 2 }),
      major: 1,
      expect: false,
      reason: "signature does not match",
    },
    {
      label: "a key for another major",
      key: mint({ email: "buyer@example.com", major: 2, privateKey }),
      major: 1,
      expect: false,
      reason: "key covers 2.x, this build is 1.x",
    },
    {
      label: "an unknown prefix",
      key: genuine.replace(/^bas1\./, "bas2."),
      major: 1,
      expect: false,
      reason: "unknown key format 'bas2'",
    },
    { label: "an empty key", key: "", major: 1, expect: false, reason: "no licence key" },
  ]),
);
